src/useAsnyc.js

```diff
@@ -4,42 +4,107 @@ import * as React from 'react';
 import usePromise from './utils/usePromise';
 
 class AsyncStore {
-  constructor(result, error, state) {
-    this.result = result;
+  constructor(result, error, state, latestValue) {
+    this._result = result;
     this._error = error;
-    this.state = state;
+    this._state = state;
+    this._latestValue = latestValue;
+  }
+
+  hasValue() {
+    return this._latestValue != null;
+  }
+
+  hasError() {
+    return this._error != null;
+  }
+
+  isLoading() {
+    return this._state === 'loading';
+  }
+
+  isSuccess() {
+    return this._state === 'success';
+  }
+
+  isFailure() {
+    return this._state === 'failure';
+  }
+
+  isInitial() {
+    return this._state === 'initial';
   }
 
   value() {
-    return this.result || null;
+    return this._latestValue ? this._latestValue.inner : undefined;
   }
 
   error() {
-    return this._error || null;
+    return this._error ? this._error : undefined;
   }
 
   success() {
-    return this.state === 'resolved' ? this.value() : null;
+    return this.isSuccess() ? this.value() : null;
   }
 
   failure() {
-    return this.state === 'rejected' ? this.error() : null;
+    return this.isFailure() ? this.error() : null;
   }
 
   initial() {
-    return this.state === 'initial';
+    return this.isInitial();
   }
 
   loading() {
-    return this.state === 'pending';
+    return this.isLoading();
+  }
+
+  match(matcher) {
+    for (const key of Object.keys(matcher)) {
+      if (key === 'value') {
+        if (this.hasValue()) {
+          return matcher.value(this.value());
+        }
+      }
+
+      if (key === 'error') {
+        if (this.hasError()) {
+          return matcher.error(this.error());
+        }
+      }
+
+      if (key === 'loading') {
+        if (this.isLoading()) {
+          return matcher.loading();
+        }
+      }
+
+      if (key === 'success') {
+        if (this.isSuccess()) {
+          return matcher.success(this.value());
+        }
+      }
+
+      if (key === 'failure') {
+        if (this.isFailure()) {
+          return matcher.failure(this.error());
+        }
+      }
+
+      if (key === 'initial') {
+        if (this.isInitial()) {
+          return matcher.initial();
+        }
+      }
+    }
```

Comment:
After the loop, check if there's a 'default' property. If so, call it with no arguments and return the result. If it doesn't exist, then we should throw a helpful error.